File: app/src/test/java/me/raatiniemi/worker/features/projects/model/ProjectsItemGetHelpTextForClockActivityAtTest.java
/*
 * Copyright (C) 2017 Worker Project
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package me.raatiniemi.worker.features.projects.model;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;
import org.junit.runners.Parameterized.Parameters;

import java.util.Arrays;
import java.util.Collection;

import me.raatiniemi.worker.domain.model.Project;

import static junit.framework.Assert.assertEquals;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@RunWith(Parameterized.class)
public class ProjectsItemGetHelpTextForClockActivityAtTest extends ProjectsItemResourceTest {
    private final String expected;
    private final Project project;

    public ProjectsItemGetHelpTextForClockActivityAtTest(
            String expected,
            Project project
    ) {
        this.expected = expected;
        this.project = project;
    }

    @Parameters
    public static Collection<Object[]> getParameters() {
        return Arrays.asList(
                new Object[][]{
                        {
                                "Clock in %s at",
                                mockProjectWithActiveIndicator(Boolean.FALSE)
                        },
                        {
                                "Clock out %s at",
                                mockProjectWithActiveIndicator(Boolean.TRUE)
                        }
                }
        );
    }

    private static Project mockProjectWithActiveIndicator(boolean isProjectActive) {
        Project project = mock(Project.class);
        when(project.getName()).thenReturn("project #1");
        when(project.isActive()).thenReturn(isProjectActive);

        return project;
    }

    @Test
    public void getHelpTextForClockActivityAt() {
        ProjectsItem projectsItem = new ProjectsItem(project);

        assertEquals(expected, projectsItem.getHelpTextForClockActivityAt(getResources()));
    }
}

File: app/src/test/java/me/raatiniemi/worker/features/projects/model/ProjectsItemGetHelpTextForClockActivityAtTest.kt
/*
 * Copyright (C) 2017 Worker Project
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package me.raatiniemi.worker.features.projects.model

import me.raatiniemi.worker.domain.model.Project
import me.raatiniemi.worker.domain.model.TimeInterval
import me.raatiniemi.worker.factory.TimeIntervalFactory
import org.junit.Assert.assertEquals
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized
import org.junit.runners.Parameterized.Parameters
import java.util.*

@RunWith(Parameterized::class)
class ProjectsItemGetHelpTextForClockActivityAtTest(
        private val expected: String,
        private val project: Project
) : ProjectsItemResourceTest() {
    @Test
    fun getHelpTextForClockActivityAt() {
        val projectsItem = ProjectsItem.from(project)

        assertEquals(expected, projectsItem.getHelpTextForClockActivityAt(resources))
    }

    companion object {
        @JvmStatic
        val parameters: Collection<Array<Any>>
            @Parameters
            get() = listOf(
                    arrayOf(
                            "Clock in %s at",
                            buildProjectWithActiveIndicator(false)
                    ),
                    arrayOf(
                            "Clock out %s at",
                            buildProjectWithActiveIndicator(true)
                    )
            )

        private fun buildProjectWithActiveIndicator(isProjectActive: Boolean): Project {
            val project = Project.builder("Project #1").build()

            if (isProjectActive) {
                val timeIntervals = ArrayList<TimeInterval>()
                timeIntervals.add(
                        TimeIntervalFactory.builder(1L)
                                .startInMilliseconds(1)
                                .stopInMilliseconds(0)
                                .build()
                )
                project.addTime(timeIntervals)
            }

            return project
        }
    }
}
